rs/replicated_state/src/canister_state/system_state/log_memory_store/index_table.rs

```diff
@@ -0,0 +1,568 @@
+    }
+}
+
+/// Index table used to speed up searches in the log’s data region.
+///
+/// The data region is a large ring buffer that stores log records. Because only
+/// the head and tail positions are known, scanning the buffer to locate a
+/// specific record would normally require reading many records sequentially.
+///
+/// The index table solves this by dividing the data region into fixed-size
+/// segments. Each segment has an `IndexEntry` pointing to the most recent record
+/// written in that segment. The table also tracks the position of the oldest
+/// live record (`front`).
+///
+/// Together, these entries let the system quickly determine which parts of the
+/// ring buffer may contain records of interest, reducing the amount of data
+/// that must be read during large or targeted searches.
+#[derive(Debug)]
+pub struct IndexTable {
+    front: Option<IndexEntry>,   // Position of the oldest live log record.
+    entries: Vec<IndexEntry>,    // Array of entries covering all data region segments.
+    segment_size: MemorySize,    // Size of each data region segment in bytes.
+    result_max_size: MemorySize, // Maximum size of results to return.
+    data_capacity: MemorySize,   // Total capacity of the data region.
+}
+
+impl IndexTable {
+    /// Creates a table that partitions the data region into fixed-size segments.
+    /// If no entries are provided, all segments start out invalid.
+    pub fn new(
+        front: Option<IndexEntry>, // front might be empty if there are no log records yet.
+        data_capacity: MemorySize,
+        index_table_pages: u16,
+        result_max_size: MemorySize,
+        entries: Vec<IndexEntry>,
+    ) -> Self {
+        let total_size_max = index_table_pages as usize * PAGE_SIZE;
+        let entry_size = INDEX_ENTRY_SIZE.get() as usize;
+        debug_assert!(entry_size > 0);
+        let entries_count = total_size_max / entry_size;
+        debug_assert!(entries_count > 0);
+        let segment_size = data_capacity.get() as usize / entries_count;
+        debug_assert!(entries_count * entry_size <= total_size_max);
+
+        let entries = if entries.is_empty() {
+            vec![IndexEntry::invalid(); entries_count]
+        } else {
+            entries
+        };
+        Self {
+            front,
+            entries,
+            segment_size: MemorySize::new(segment_size as u64),
+            result_max_size,
+            data_capacity,
+        }
+    }
+
+    /// Records the most recent log entry associated with the segment that contains `position`.
+    ///
+    /// This operation assumes that log records arrive strictly in increasing `idx` order —
+    /// meaning each update corresponds to a newer record than any previously stored one.
+    /// Because segments keep only the newest record, providing an out-of-order record would
+    /// silently overwrite newer data and corrupt segment ordering.
+    ///
+    /// IMPORTANT: callers must guarantee monotonic `idx` progression.
```

Comment:
The timestamps also need to be monotonic, right? Should we check that in `append_log` when we're also checking the  index order?